File: src/com/train/algorithm/dynamicProgramming/implementInC++/UniquePaths.cpp
//
// Created by 罗斯 on 27/11/2022.
//
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int uniquePaths(int m, int n) {
//        if (m < 0 || n < 0) return 0;
//        if (m == 1 && n == 1) return 1;
//        if (f[m][n] > 0) return f[m][n];
//        int left = uniquePaths(m - 1, n);
//        int top = uniquePaths(m, n - 1);
//        f[m][n] = left + top;
//        return f[m][n];

        auto f = vector<vector<int>>(n + 1, vector<int>(m + 1, 0));
        f[1][1] = 1;

        for (int y = 1; y <= n; ++y)
            for(int x = 1; x <= m; ++x) {
                if (x == 1 && y == 1) {
                    continue;
                } else {
                    f[y][x] = f[y-1][x] + f[y][x-1];
                }
            }

        return f[n][m];
    }
private:
    unordered_map<int, unordered_map<int, int>> f;
};

File: src/com/train/algorithm/dynamicProgramming/implementInPython/UniquePaths.py
class Solution(object):
    def uniquePaths(self, m, n):
        matrix = [[0] * m for i in range(n)]
        matrix[0][0] = 1
        for i in range(1, m):
            matrix[0][i] = matrix[0][i-1]
        for j in range(1, n):
            matrix[j][0] = matrix[j-1][0]
        for i in range(1, n):
            for j in range(1, m):
                matrix[i][j] = matrix[i-1][j] + matrix[i][j - 1]
        return matrix[n-1][m-1]